feat: circular autofix/conflicting rules detection

Adrastopoulos opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [OpenJS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain: #17609

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)

This addresses an issue with ESLint's autofix mechanism, where it could enter a loop in cases of conflicting rules/circular fixes (e.g., alternating fixes that add and remove the same text).

The fix ensures that the verifyAndFix method detects circular fixes by comparing the output of the current fix pass to the output two passes ago. If a circular pattern is detected, the autofix process halts and shows a warning.

Both the Flat Config and legacy Config tests are updated to handle this.

Fixes #17609.

#### Is there anything you'd like reviewers to focus on?

1. Review how the circular fix warning is handled. I implemented it using `process.emitWarning` similar to [@snitin315's suggestion](https://github.com/eslint/eslint/issues/17609#issuecomment-2412979714). This includes emitting a warning and adding a `debug` log message. Let me know if this aligns with ESLint's best practices.  

2. Verify the new test cases added under the `verifyAndFix` namespace for both Flat Config (`FlatConfigArray`) and legacy Config setups.

<!-- markdownlint-disable-file MD004 -->

mdjermanovic commented on lib/linter/linter.js:
```diff
@@ -2340,6 +2341,23 @@ class Linter {
+            // Stop if we've made a circular fix
+            if (recentOutputs.length === 2 && fixedResult.output === recentOutputs[0]) {
+                debug(`Circular fixes detected after pass ${passNumber}. Exiting fix loop.`);
+                process.emitWarning(
+                    "Warning: Circular fixes detected in your configuration.",
+                    "ESLINT_CIRCULAR_FIXES"
+                );
```
The `Linter` class can be used in browsers, where `process` global doesn't exist. That's why the browser test is failing:

```
[0-0] TypeError in "Linter.verifyAndFix.should stop fixing if a circular fix is detected"
TypeError: n.emitWarning is not a function
    at Linter.verifyAndFix (http://localhost:45903/build/eslint.js:2:665704)
    at Context.<anonymous> (http://localhost:45903/home/runner/work/eslint/eslint/tests/lib/linter/linter.js:7621:38)
[0-0] TypeError in "Linter with FlatConfigArray.verifyAndFix().should stop fixing if a circular fix is detected"
TypeError: n.emitWarning is not a function
    at Linter.verifyAndFix (http://localhost:45903/build/eslint.js:2:665704)
    at Context.<anonymous> (http://localhost:45903/home/runner/work/eslint/eslint/tests/lib/linter/linter.js:16[50](https://github.com/eslint/eslint/actions/runs/12254524290/job/34185621473#step:5:51)1:38)
```

If we want to provide a warning using `process.emitWarning()`, we should check if `process` exists.

Though, I'm not sure if we clarified whether we want to use `process.emitWarning` or add a warning lint message. @nzakas what do you think?

nzakas replied:
It's not technically a lint warning message, so I don't think it should affect things like `--max-warnings`.

I'm in favor of using `process.emitWarning()` if `process` exists. Maybe just:

```js
globalThis.process?.emitWarning("...");
```

mdjermanovic replied:
Makes sense to me. To be even safer, maybe we could also add `?.` after `emitWarning`. And I think we should include the filename in the warning message.

So, perhaps something like this:

```js
globalThis.process?.emitWarning?.(
    `Circular fixes detected while fixing ${options?.filename ?? "text"}. It is likely that you have conflicting rules in your configuration.`,
    "ESLintCircularFixesWarning"
);
```

nzakas commented on lib/linter/linter.js:
```diff
@@ -2340,6 +2341,23 @@ class Linter {
             }
 
+            // Stop if we've made a circular fix
+            if (recentOutputs.length === 2 && fixedResult.output === recentOutputs[0]) {
```
Because it's expensive to compare large strings, let's check the length first to make sure it's worth taking that performance hit.
```suggestion
            if (recentOutputs.length === 2 && fixedResult.output.length === recentOutputs[0].length && fixedResult.output === recentOutputs[0]) {
```

cirex-web replied:
Is `fixedResult.output` long enough so that this is a significant performance optimization?

nzakas replied:
Even comparing single-character strings in a loop can cause a significant performance degradation.

maxpatiiuk replied:
to reduce performance penalty, would it make sense to move this logic inside the `if (fixedResult.fixed) {` (line 2391)?
in that case, you can check for `passNumber === MAX_AUTOFIX_PASSES` before checking autofix results

cirex-web commented:
Have you considered (or should we consider) cycles of length greater than 2?

Adrastopoulos commented:
> Have you considered (or should we consider) cycles of length greater than 2? (why not just keep a set of all previously seen results? (or the hash of every result, if the lengths are too long))

If this is the case, we could use a set to track all previous results rather than only checking `n` vs `n-2`, terminating when the set size does not increase.

nzakas commented:
Keeping a set of all previously seen results increases memory usage. I don't think there's a good reason to keep more than two based on our discussion in the original issue.

mdjermanovic commented on lib/linter/linter.js:
```diff
@@ -2290,6 +2290,7 @@ class Linter {
         const debugTextDescription = options && options.filename || `${text.slice(0, 10)}...`;
         const shouldFix = options && typeof options.fix !== "undefined" ? options.fix : true;
         const stats = options?.stats;
+        const recentOutputs = [];
```
I think the initial text should be in the list so that the autofixing stops if the second pass produces output that is the same as the initial text.

Adrastopoulos replied:
Correct me if I'm wrong, but my understanding is that we are providing some initial state by virtue of 
https://github.com/Adrastopoulos/eslint/blob/fix/circular-autofix-detection/tests/lib/linter/linter.js#L16559

mdjermanovic replied:
Yes, the initial state is the original source code, but would it be detected that the output of the second fix pass is equal to the original source code?

mdjermanovic commented on tests/lib/linter/linter.js:
```diff
@@ -7590,6 +7590,44 @@ var a = "test2";
+            const suppressedMessages = linter.getSuppressedMessages();
+
+            assert.strictEqual(suppressedMessages.length, 0, "No suppressed messages should exist.");
+        });
```
In these tests, can we also check if the warning was emitted?

Adrastopoulos replied:
How can we do this?

mdjermanovic commented on tests/lib/linter/linter.js:
```diff
@@ -7590,6 +7590,44 @@ var a = "test2";
+            assert.strictEqual(fixResult.fixed, true, "Fixing should have been applied but stopped due to circular fixes.");
+            assert.strictEqual(fixResult.output, "-a", "Output should match the original input due to circular fixes.");
+            assert.strictEqual(fixResult.messages.length, 1, "There should be one remaining lint message after detecting circular fixes.");
+            assert.strictEqual(fixResult.messages[0].message, "Circular fix detected", "Message should match the last fix attempt.");
```
I think there should be different messages for adding and removing a hyphen as this doesn't really ensure that the message is correct for the final output.

Adrastopoulos replied:
Good point. I've pushed some changes. Let me know your thoughts.

Tanujkanti4441 commented:
Hi @Adrastopoulos, there are some suggestions you can check out.

nzakas commented:
@Adrastopoulos are you still working on this?

Adrastopoulos commented:
> @Adrastopoulos are you still working on this?

I'll have some time to look at this this week. School started up and things have been hectic.

nzakas commented:
Sounds good. Let us know if you'd like help finishing this up.

Adrastopoulos commented:
I've addressed the feedback now. Assuming nothing else comes up, it will suffice to check if the warning was emitted in text to complete this PR. I am not sure how to implement that. @mdjermanovic any pointers?

nzakas commented:
Take a look at https://github.com/eslint/eslint/blob/main/tests/lib/cli.js. There are multiple tests in there that check for output from the CLI.

Adrastopoulos commented:
> Take a look at https://github.com/eslint/eslint/blob/main/tests/lib/cli.js. There are multiple tests in there that check for output from the CLI.

I am stuck. I created a stub for the emitWarning method akin to `cli.js`:

```ts
let processStub;

beforeEach(() => {
    sinon.restore();
    processStub = sinon.stub(process, "emitWarning");
});

afterEach(() => {
    processStub.restore();
});
```

But this stub is never being called in the test.

mdjermanovic commented:
> I am stuck. I created a stub for the emitWarning method akin to `cli.js`:
> 
> ```ts
> let processStub;
> 
> beforeEach(() => {
>     sinon.restore();
>     processStub = sinon.stub(process, "emitWarning");
> });
> 
> afterEach(() => {
>     processStub.restore();
> });
> ```
> 
> But this stub is never being called in the test.

Can you push the changes so we can figure out what's causing the problem?

Adrastopoulos commented:
> Can you push the changes so we can figure out what's causing the problem?

Done.

mdjermanovic commented on tests/lib/linter/linter.js:
```diff
@@ -7590,6 +7590,76 @@ var a = "test2";
+                                        return;
+                                    }
+
+                                    fixer.insertTextBefore(node, "-");
```
```suggestion
                                    return fixer.insertTextBefore(node, "-");
```

mdjermanovic commented on tests/lib/linter/linter.js:
```diff
@@ -7590,6 +7590,76 @@ var a = "test2";
+                                        return;
+                                    }
+
+                                    fixer.removeRange([0, 1]);
```
```suggestion
                                    return fixer.removeRange([0, 1]);
```

mdjermanovic commented on tests/lib/linter/linter.js:
```diff
@@ -16421,6 +16502,91 @@ var a = "test2";
+                                                        return;
+                                                    }
+
+                                                    fixer.insertTextBefore(node, "-");
```
```suggestion
                                                    return fixer.insertTextBefore(node, "-");
```

mdjermanovic commented on tests/lib/linter/linter.js:
```diff
@@ -16421,6 +16502,91 @@ var a = "test2";
+                                                        return;
+                                                    }
+
+                                                    fixer.removeRange([0, 1]);
```
```suggestion
                                                    return fixer.removeRange([0, 1]);
```

mdjermanovic reviewed (commented):
I believe the problem is that the `fix` methods do not return any fixes, so the test case doesn't really produce the expected warning.

nzakas commented:
@Adrastopoulos there are a few more comments. Can you take a look?

nzakas commented:
@Adrastopoulos would you like someone else to finish this work up?

nzakas commented:
No worries. @mdjermanovic do you want to pick this up?

mdjermanovic commented:
Yes, I'll take this.

mdjermanovic commented:
Continued in https://github.com/eslint/eslint/pull/19514.
